test/common/http/header_map_impl_test.cc

```diff
@@ -319,15 +319,67 @@ TEST(HeaderMapImplTest, DoubleInlineAdd) {
   EXPECT_EQ(1UL, headers.size());
 }
 
+TEST(HeaderMapImplTest, AddCopy) {
+  HeaderMapImpl headers;
+
+  // Start with a string value.
+  std::unique_ptr<LowerCaseString> lcKeyPtr(new LowerCaseString("hello"));
+  headers.addCopy(*lcKeyPtr, "world");
+
+  const HeaderString& value = headers.get(*lcKeyPtr)->value();
+
+  EXPECT_STREQ("world", value.c_str());
+  EXPECT_EQ(5UL, value.size());
+
+  lcKeyPtr.reset(nullptr);
+
+  const HeaderString& value2 = headers.get(LowerCaseString("hello"))->value();
+
+  EXPECT_STREQ("world", value2.c_str());
+  EXPECT_EQ(5UL, value2.size());
+  EXPECT_EQ(value.c_str(), value2.c_str());
+  EXPECT_EQ(1UL, headers.size());
+
+  // Repeat with an int value.
+  //
+  // @kflynn: addReferenceKey and addCopy can both add multiple instances of a
+  // given header, so we need to delete the old "hello" header.
+  headers.remove(LowerCaseString("hello"));
+
+  lcKeyPtr.reset(new LowerCaseString(std::string("he") + "llo")); // Try to force the compiler's hand here.
+
+  headers.addCopy(*lcKeyPtr, 42);
+
+  const HeaderString& value3 = headers.get(*lcKeyPtr)->value();
+
+  EXPECT_STREQ("42", value3.c_str());
+  EXPECT_EQ(2UL, value3.size());
+
+  lcKeyPtr.reset(nullptr);
+
+  const HeaderString& value4 = headers.get(LowerCaseString("hello"))->value();
+
+  EXPECT_STREQ("42", value4.c_str());
+  EXPECT_EQ(2UL, value4.size());
+  EXPECT_EQ(1UL, headers.size());
+
+  // We can make yet another different key to look up here.
+  LowerCaseString lcKey3(std::string("he") + "ll" + "o"); // Try to force the compiler's hand here.
```

Comment:
Please remove trailing comment about compiler. Not needed (and not sure exactly what it means).